<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
    "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<meta charset="utf-8">
<title>Issue 2093: Throws clause of condition_variable::wait with predicate</title>
<style type="text/css">
  p {text-align:justify}
  li {text-align:justify}
  blockquote.note
  {
    background-color:#E0E0E0;
    padding-left: 15px;
    padding-right: 15px;
    padding-top: 1px;
    padding-bottom: 1px;
  }
  ins {background-color:#A0FFA0}
  del {background-color:#FFA0A0}
  table {border-collapse: collapse;}
</style>
</head>
<body>
<hr>
<h3><a name="2093" href="lwg-defects.html#2093">2093.</a> Throws clause of <tt>condition_variable::wait</tt> with predicate</h3>
<p><b>Section:</b> 33.5.3 <a href="https://timsong-cpp.github.io/cppwp/thread.condition.condvar">[thread.condition.condvar]</a> <b>Status:</b> <a href="lwg-active.html#C++14">C++14</a>
 <b>Submitter:</b> Alberto Ganesh Barbati <b>Opened:</b> 2011-10-27 <b>Last modified:</b> 2015-10-20 02:52:46 UTC</p>
<p><b>Priority: </b>Not Prioritized
</p>
<p><b>View all other</b> <a href="lwg-index.html#thread.condition.condvar">issues</a> in <a href="https://timsong-cpp.github.io/cppwp/thread.condition.condvar">[thread.condition.condvar]</a>.</p>
<p><b>View all issues with</b> <a href="lwg-status.html#C++14">C++14</a> status.</p>
<p><b>Discussion:</b></p>

<p>
the Throws: clause of <tt>condition_variable::wait&#47;wait_xxx</tt> functions that 
take a predicate argument is:
</p>
<blockquote><p>
<i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>).
</p></blockquote>
<p>
If executing the predicate throws an exception, I would expect such exception to propagate unchanged 
to the caller, but the throws clause seems to indicate that it gets mutated into a <tt>system_error</tt>.
That's because of 20.4.1.4 <a href="https://timsong-cpp.github.io/cppwp/structure.specifications">[structure.specifications]</a>&#47;4:
<p/>
"If <tt>F</tt>'s semantics contains a Throws:, Postconditions:, or Complexity: element, then that supersedes 
any occurrences of that element in the code sequence."
<p/>
Is my interpretation correct? Does it match the intent?
<p/>
Daniel comments:
<p/>
I don't think that this interpretation is entirely correct, the wording does not say that 
<tt>std::system_error</tt> or a derived class must be thrown, it simply is underspecified 
in this regard (The extreme interpretation is that the behaviour would be undefined, but 
that would be too far reaching I think). We have better wording for this in 
33.4.6.2 <a href="https://timsong-cpp.github.io/cppwp/thread.once.callonce">[thread.once.callonce]</a> p4, where it says:
<p/>
"<i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>), 
or any exception thrown by <tt>func</tt>."
<p/>
or in 33.3.3 <a href="https://timsong-cpp.github.io/cppwp/thread.thread.this">[thread.thread.this]</a> p6&#47;p9:
<p/>
"<i>Throws</i>: Nothing if <tt>Clock</tt> satisfies the <tt>TrivialClock</tt> requirements 
(23.17.3 <a href="https://timsong-cpp.github.io/cppwp/time.clock.req">[time.clock.req]</a>) and operations of <tt>Duration</tt> do not throw exceptions. 
[ <i>Note</i>: instantiations of time point types and clocks supplied by the implementation 
as specified in 23.17.7 <a href="https://timsong-cpp.github.io/cppwp/time.clock">[time.clock]</a> do not throw exceptions. &mdash; <i>end note</i> ]"
<p/>
So, the here discussed Throws elements should add lines along the lines of
<p/>
"Any exception thrown by operations of <tt>pred</tt>."
<p/>
and similar wording for time-related operations:
<p/>
"Any exception thrown by operations of <tt>Duration</tt>",
<p/>
"Any exception thrown by operations of <tt>chrono::duration&lt;Rep, Period&gt;</tt>"
</p>

<p><i>[2011-11-28: Ganesh comments and suggests wording]</i></p>


<p>
As for the discussion about the exception thrown by the manipulation of time-related objects, 
I believe the argument applies to all functions declared in 33 <a href="https://timsong-cpp.github.io/cppwp/thread">[thread]</a>. Therefore, 
instead of adding wording to each member, I would simply move those requirements from 
33.3.3 <a href="https://timsong-cpp.github.io/cppwp/thread.thread.this">[thread.thread.this]</a> p6&#47;p9 to a new paragraph in 33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>. 
<p/>
As for 33.5.4 <a href="https://timsong-cpp.github.io/cppwp/thread.condition.condvarany">[thread.condition.condvarany]</a>, the member functions <tt>wait()</tt> and 
<tt>wait_until()</tt> are described only in terms of the Effects: clause (so strictly speaking, 
they need no changes), however, <tt>wait_for()</tt> is described with a full set of clauses 
including Throws: and Error conditions:. Either we should add those clauses to <tt>wait&#47;wait_until</tt> 
with changes similar to the one above, or remove paragraphs 29 to 34 entirely. By the way, 
even paragraph 26 could be removed IMHO.
</p>

<p><i>[2012, Kona]</i></p>

<p>
We like the idea behind the proposed resolution.
<p/>
Modify the first addition to read instead: "Functions that specify a timeout, will throw if an operation 
on a clock, time point, or time duration throws an exception." 
<p/>
In the note near the bottom change "even if the timeout has already expired" to "or if the timeout has 
already expired". (This is independent, but the original doesn't seem to make sense.) 
<p/>
Move to Review.
</p>

<p><i>[2012, Portland]</i></p>

<p>
Concurrency move to Ready with slightly ammended wording.
</p>

<p><i>[2013-04-20 Bristol]</i></p>



<p><b>Proposed resolution:</b></p>
<p>This wording is relative to N3337.</p>

<ol>
<li><p>Add a new paragraph at the end of 33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>:</p>

<blockquote>
<p>
[&hellip;]
<p/>
-6- The resolution of timing provided by an implementation depends on both operating system and hardware.
The finest resolution provided by an implementation is called the <i>native resolution</i>.
<p/>
-7- Implementation-provided clocks that are used for these functions shall meet the <tt>TrivialClock</tt> 
requirements (23.17.3 <a href="https://timsong-cpp.github.io/cppwp/time.clock.req">[time.clock.req]</a>).
<p/>

<ins>-?- Functions that specify a timeout, will throw if, during the execution of this function,
a clock, time point, or time duration throws an exception. [ <i>Note</i>: instantiations of clock,
time point and duration types supplied by the implementation as specified in 23.17.7 <a href="https://timsong-cpp.github.io/cppwp/time.clock">[time.clock]</a>
do not throw exceptions. &mdash; <i>end note</i>]</ins>
</p>
</blockquote>
</li>


<li><p>Change 33.3.3 <a href="https://timsong-cpp.github.io/cppwp/thread.thread.this">[thread.thread.this]</a> as indicated:</p>

<blockquote><pre>
template &lt;class Clock, class Duration&gt;
  void sleep_until(const chrono::time_point&lt;Clock, Duration&gt;&amp; abs_time);;
</pre><blockquote>
<p>
-4- <i>Effects</i>: Blocks the calling thread for the absolute timeout (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>) specified by <tt>abs_time</tt>.
<p/>
-5- <i>Synchronization</i>: None.
<p/>
-6- <i>Throws</i>: <ins>timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>).</ins><del>Nothing if 
<tt>Clock</tt> satisfies the <tt>TrivialClock</tt> requirements (23.17.3 <a href="https://timsong-cpp.github.io/cppwp/time.clock.req">[time.clock.req]</a>) and 
operations of <tt>Duration</tt> do not throw exceptions. [ <i>Note</i>: instantiations of time point types 
and clocks supplied by the implementation as specified in 23.17.7 <a href="https://timsong-cpp.github.io/cppwp/time.clock">[time.clock]</a> do not throw 
exceptions. &mdash; <i>end note</i>]</del>
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Rep, class Period&gt;
  void sleep_for(const chrono::duration&lt;Rep, Period&gt;&amp; rel_time);;
</pre><blockquote>
<p>
-7- <i>Effects</i>: Blocks the calling thread for the relative timeout (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>) specified by <tt>rel_time</tt>.
<p/>
-8- <i>Synchronization</i>: None.
<p/>
-9- <i>Throws</i>: <ins>timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>).</ins><del>Nothing 
if operations of <tt>chrono::duration&lt;Rep, Period&gt;</tt> do not throw exceptions. 
[ <i>Note</i>: instantiations of time point types and clocks supplied by the implementation
as specified in 23.17.7 <a href="https://timsong-cpp.github.io/cppwp/time.clock">[time.clock]</a> do not throw exceptions. &mdash; <i>end note</i>]</del>
</p>
</blockquote></blockquote>

</li>

<li><p>Change 33.4.3.3 <a href="https://timsong-cpp.github.io/cppwp/thread.timedmutex.requirements">[thread.timedmutex.requirements]</a> as indicated:</p>

<p>
-3- The expression <tt>m.try_lock_for(rel_time)</tt> shall be well-formed and have the following semantics:
<p/>
[&hellip;]
<p/>
-5- <i>Effects</i>: The function attempts to obtain ownership of the mutex within the relative timeout (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>)
specified by <tt>rel_time</tt>. If the time specified by <tt>rel_time</tt> is less than or equal to <tt>rel_time.zero()</tt>, the
function attempts to obtain ownership without blocking (as if by calling <tt>try_lock()</tt>). The function
shall return within the timeout specified by <tt>rel_time</tt> only if it has obtained ownership of the mutex
object. [<i>Note</i>: As with <tt>try_lock()</tt>, there is no guarantee that ownership will be obtained if the lock
is available, but implementations are expected to make a strong effort to do so. &mdash; <i>end note</i>]
<p/>
[&hellip;]
<p/>
-8- <i>Synchronization</i>: If <tt>try_lock_for()</tt> returns <tt>true</tt>, prior <tt>unlock()</tt> operations on the same object
<i>synchronize with</i> (4.7 <a href="https://timsong-cpp.github.io/cppwp/intro.multithread">[intro.multithread]</a>) this operation.
<p/>
-9- <i>Throws</i>: <ins>timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>).</ins><del>Nothing</del>.
<p/>
-10- The expression <tt>m.try_lock_until(abs_time)</tt> shall be well-formed and have the following semantics:
<p/>
[&hellip;]
<p/>
-12- <i>Effects</i>: The function attempts to obtain ownership of the mutex. If <tt>abs_time</tt> has already passed, the
function attempts to obtain ownership without blocking (as if by calling <tt>try_lock()</tt>). The function
shall return before the absolute timeout (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>) specified by <tt>abs_time</tt> only 
if it has obtained ownership of the mutex object. [<i>Note</i>: As with <tt>try_lock()</tt>, there is no guarantee 
that ownership will be obtained if the lock is available, but implementations are expected to make a strong effort 
to do so. &mdash; <i>end note</i>]
<p/>
[&hellip;]
<p/>
-15- <i>Synchronization</i>: If <tt>try_lock_until()</tt> returns true, prior <tt>unlock()</tt> operations on the same object
<i>synchronize with</i> (4.7 <a href="https://timsong-cpp.github.io/cppwp/intro.multithread">[intro.multithread]</a>) this operation.
<p/>
-16- <i>Throws</i>: <ins>timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>).</ins><del>Nothing</del>.
</p>
</li>

<li><p>Change 33.5.3 <a href="https://timsong-cpp.github.io/cppwp/thread.condition.condvar">[thread.condition.condvar]</a> as indicated:</p>

<blockquote><pre>
template &lt;class Predicate&gt;
  void wait(unique_lock&lt;mutex&gt;&amp; lock, Predicate pred);
</pre><blockquote>
<p>
[&hellip;]
<p/>
-15- <i>Effects</i>: <ins>Equivalent to:</ins>
</p>
<blockquote><pre>
while (!pred())
  wait(lock);
</pre></blockquote>
<p>
[&hellip;]
<p/>
-17- <i>Throws</i>: <tt>std::system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>)<ins>, 
timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>), or any exception thrown by <tt>pred</tt></ins>.
<p/>
[&hellip;]
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Clock, class Duration&gt;
  cv_status wait_until(unique_lock&lt;mutex&gt;&amp; lock,
                       const chrono::time_point&lt;Clock, Duration&gt;&amp; abs_time);
</pre><blockquote>
<p>
[&hellip;]
<p/>
-23- <i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>)
<ins>or timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>)</ins>.
<p/>
[&hellip;]
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Rep, class Period&gt;
  cv_status wait_for(unique_lock&lt;mutex&gt;&amp; lock,
                     const chrono::duration&lt;Rep, Period&gt;&amp; rel_time);
</pre><blockquote>
<p>
[&hellip;]
<p/>
-26- <i>Effects</i>: <del>as if</del><ins>Equivalent to:</ins>
</p>
<blockquote><pre>
return wait_until(lock, chrono::steady_clock::now() + rel_time);
</pre></blockquote>
<p>
[&hellip;]
<p/>
-29- <i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>)
<ins>or timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>)</ins>.
<p/>
[&hellip;]
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Clock, class Duration, class Predicate&gt;
  bool wait_until(unique_lock&lt;mutex&gt;&amp; lock,
                  const chrono::time_point&lt;Clock, Duration&gt;&amp; abs_time,
                  Predicate pred);
</pre><blockquote>
<p>
[&hellip;]
<p/>
-32- <i>Effects</i>: <ins>Equivalent to:</ins>
</p>
<blockquote><pre>
while (!pred())
  if (wait_until(lock, abs_time) == cv_status::timeout)
    return pred();
return true;
</pre></blockquote>
<p>
<del>-33- <i>Returns</i>: <tt>pred()</tt></del>
<p/>
[&hellip;]
<p/>
-36- <i>Throws</i>: <tt>std::system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>)<ins>, 
timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>), or any exception thrown by <tt>pred</tt></ins>.
<p/>
[&hellip;]
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Rep, class Period, class Predicate&gt;
  bool wait_for(unique_lock&lt;mutex&gt;&amp; lock,
                const chrono::duration&lt;Rep, Period&gt;&amp; rel_time,
                Predicate pred);
</pre><blockquote>
<p>
[&hellip;]
<p/>
-39- <i>Effects</i>: <del>as if</del><ins>Equivalent to:</ins>
</p>
<blockquote><pre>
return wait_until(lock, chrono::steady_clock::now() + rel_time, std::move(pred));
</pre></blockquote>
<p>
[&hellip;]
<p/>
<del>-42- <i>Returns</i>: <tt>pred()</tt></del>
<p/>
[&hellip;]
<p/>
-44- <i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>)<ins>, 
timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>), or any exception thrown by <tt>pred</tt></ins>.
<p/>
[&hellip;]
</p>
</blockquote></blockquote>

</li>

<li><p>Change 33.5.4 <a href="https://timsong-cpp.github.io/cppwp/thread.condition.condvarany">[thread.condition.condvarany]</a> as indicated:</p>

<blockquote><pre>
template &lt;class Lock, class Predicate&gt;
  void wait(Lock&amp; lock, Predicate pred);
</pre><blockquote>
<p>
-14- <i>Effects</i>: <ins>Equivalent to:</ins>
</p>
<blockquote><pre>
while (!pred())
  wait(lock);
</pre></blockquote>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Lock, class Clock, class Duration&gt;
  cv_status wait_until(Lock&amp; lock, const chrono::time_point&lt;Clock, Duration&gt;&amp; abs_time);
</pre><blockquote>
<p>
[&hellip;]
<p/>
-18- <i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>)
<ins>or any timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>)</ins>.
<p/>
[&hellip;]
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Lock, class Rep, class Period&gt;
  cv_status wait_for(Lock&amp; lock, const chrono::duration&lt;Rep, Period&gt;&amp; rel_time);
</pre><blockquote>
<p>
[&hellip;]
<p/>
-20- <i>Effects</i>: <del>as if</del><ins>Equivalent to:</ins>
</p>
<blockquote><pre>
return wait_until(lock, chrono::steady_clock::now() + rel_time);
</pre></blockquote>
<p>
[&hellip;]
<p/>
-23- <i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>)
<ins>or any timeout-related exceptions (33.2.4 <a href="https://timsong-cpp.github.io/cppwp/thread.req.timing">[thread.req.timing]</a>)</ins>.
<p/>
[&hellip;]
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Lock, class Clock, class Duration, class Predicate&gt;
  bool wait_until(Lock&amp; lock, const chrono::time_point&lt;Clock, Duration&gt;&amp; abs_time, Predicate pred);
</pre><blockquote>
<p>
-25- <i>Effects</i>: <ins>Equivalent to:</ins>
</p>
<blockquote><pre>
while (!pred())
  if (wait_until(lock, abs_time) == cv_status::timeout)
    return pred();
return true;
</pre></blockquote>
<p>
-26- <del><i>Returns</i>: <tt>pred()</tt></del><ins>[<i>Note</i>: There is no blocking if <tt>pred()</tt> is initially <tt>true</tt>, 
or if the timeout has already expired. &mdash; <i>end note</i>]</ins>
<p/>
-27- [<i>Note</i>: The returned value indicates whether the predicate evaluates to <tt>true</tt> regardless of whether the
timeout was triggered. <i>end note</i>]
</p>
</blockquote></blockquote>

<blockquote><pre>
template &lt;class Lock, class Rep, class Period, class Predicate&gt;
  bool wait_for(Lock&amp; lock, const chrono::duration&lt;Rep, Period&gt;&amp; rel_time, Predicate pred);
</pre><blockquote>
<p>
-28- <i>Effects</i>: <del>as if</del><ins>Equivalent to:</ins>
</p>
<blockquote><pre>
return wait_until(lock, chrono::steady_clock::now() + rel_time, std::move(pred));
</pre></blockquote>
<p>
<del>-29- [<i>Note</i>: There is no blocking if <tt>pred()</tt> is initially <tt>true</tt>, 
even if the timeout has already expired. &mdash; <i>end note</i>]</del>
<del>-30- <i>Postcondition</i>: <tt>lock</tt> is locked by the calling thread.</del>
<p/>
<del>-31- <i>Returns</i>: <tt>pred()</tt></del>
<p/>
<del>-32- [<i>Note</i>: The returned value indicates whether the predicate evaluates to <tt>true</tt> 
regardless of whether the timeout was triggered. &mdash; <i>end note</i>]</del>
<p/>
<del>-33- <i>Throws</i>: <tt>system_error</tt> when an exception is required (33.2.2 <a href="https://timsong-cpp.github.io/cppwp/thread.req.exception">[thread.req.exception]</a>).</del>
<p/>
<del>-34- <i>Error conditions</i>:</del>
</p>
<ul>
<li><del>equivalent error condition from <tt>lock.lock()</tt> or <tt>lock.unlock()</tt>.</del></li>
</ul>
</blockquote></blockquote>

</li>
</ol>





</body>
</html>
